vertexbufferlayout.cpp

```diff
@@ -0,0 +1,51 @@
+#include "vertexbufferlayout.h"
+
+#include <vector>
+
+#include "renderutils.h"
+
+VertexBufferLayout::VertexBufferLayout() :
+    mStride(0)
+{
+}
+
+VertexBufferLayout::~VertexBufferLayout()
+{
+}
+
+template<typename T>
+void VertexBufferLayout::Push(unsigned int count)
+{
+    static_assert(false);
+}
+
+template<>
+void VertexBufferLayout::Push<float>(unsigned int count)
+{
+    mElements.push_back({GL_FLOAT, count, false});
+    mStride += count * sizeof(GLfloat);
+}
+
+template<>
+void VertexBufferLayout::Push<unsigned int>(unsigned int count)
+{
+    mElements.push_back({GL_UNSIGNED_INT, count, false});
+    mStride += count * sizeof(GLuint);
+}
+
+template<>
+void VertexBufferLayout::Push<unsigned char>(unsigned int count)
+{
+    mElements.push_back({GL_UNSIGNED_BYTE, count, true});
+    mStride += count * sizeof(GLubyte);
+}
+
+const std::vector<VertexBufferElement>& VertexBufferLayout::GetElements() const
```

Comment:
In general you should think about not exposing your internal member variables directly to the outside world. It breaks the abstraction of your class. We did it a ton at spatial and I think it leads to a lot of issues down the road.

Instead, consider if you can do the same operations internal to the class, or just use a vector directly instead of wrapping it in a class to begin with.